src/Universum/Monad/Either.hs

```diff
@@ -82,24 +86,46 @@ maybeToRight l = maybe (Left l) Right
 maybeToLeft :: r -> Maybe l -> Either l r
 maybeToLeft r = maybe (Right r) Left
 
--- | Applies given action to 'Either' content if 'Left' is given.
-whenLeft :: Applicative f => Either l r -> (l -> f ()) -> f ()
-whenLeft (Left  l) f = f l
-whenLeft (Right _) _ = pass
+-- | Applies given action to 'Either' content if 'Left' is given and returns
+-- the result. In case of 'Right' the default value will be returned.
+whenLeft :: Applicative f => Either l r -> a -> (l -> f a) -> f a
+whenLeft (Left  l) _ f = f l
+whenLeft (Right _) a _ = pure a
 {-# INLINE whenLeft #-}
 
+-- | Applies given action to 'Either' content if 'Left' is given.
+whenLeft_ :: Applicative f => Either l r -> (l -> f ()) -> f ()
+whenLeft_ e = whenLeft e ()
+{-# INLINE whenLeft_ #-}
+
 -- | Monadic version of 'whenLeft'.
-whenLeftM :: Monad m => m (Either l r) -> (l -> m ()) -> m ()
-whenLeftM me f = me >>= \e -> whenLeft e f
+whenLeftM :: Monad m => m (Either l r) -> a -> (l -> m a) -> m a
+whenLeftM me a f = me >>= \e -> whenLeft e a f
 {-# INLINE whenLeftM #-}
 
--- | Applies given action to 'Either' content if 'Right' is given.
-whenRight :: Applicative f => Either l r -> (r -> f ()) -> f ()
-whenRight (Left  _) _ = pass
-whenRight (Right r) f = f r
+-- | Monadic version of 'whenLeft_'.
+whenLeftM_ :: Monad m => m (Either l r) -> (l -> m ()) -> m ()
+whenLeftM_ me f = me >>= \e -> whenLeft_ e f
+{-# INLINE whenLeftM_ #-}
+
+-- | Applies given action to 'Either' content if 'Right' is given and returns
+-- the result. In case of 'Left' the default value will be returned.
+whenRight :: Applicative f => Either l r -> a -> (r -> f a) -> f a
+whenRight (Left  _) a _ = pure a
+whenRight (Right r) _ f = f r
 {-# INLINE whenRight #-}
 
+-- | Applies given action to 'Either' content if 'Right' is given.
+whenRight_ :: Applicative f => Either l r -> (r -> f ()) -> f ()
```

Comment:
Maybe you could also update `.hlint.yaml` and replace every `when[Left|Right]` by adding `_` at the end? And probably we should updates types of `whenLeft` in `README.md`...